### PR TITLE
Add support for GROUPING()

### PR DESCRIPTION
**Pull request by petroav**
- The arguments to GROUPING() are not evaluated. The function returns
  a bitmap as an integer where the rightmost column in the GROUPING()
  argument list is the least significant bit and is set to 1 when
  that column is not present in the grouping and to 0 otherwise.
- Ensure each argument to GROUPING() is a grouping column (i.e. a
  column in GROUP BY, GROUPING SETS, ROLLUP or CUBE).
- Parse GROUPING() to it's own AST node and rewrite it to a
  FunctionCall expression in QueryPlanner.
- Rewrite GROUPING() to a 0 literal when used in a GROUP BY.

### CODE REVIEW

**Comment by petroav** [comment]
Putting this up to get approval on design (location of rewrite). Once that is approved I'll add unit tests to this commit and a second commit with documentation and release notes.

@raghavsethi when you get a chance I'd appreciate it if you took a look. I know you're trying to get a release out so if you don't have time at the moment that's fine, I'll ping you again once 0.148 is out.

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/GroupingOperationFunction.java [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+
+    @ScalarFunction
+    @Description("Returns a bitmap as an integer, indicating which columns are present in the grouping")
+    @SqlType(BIGINT)
+    public static long grouping(
+            @SqlType(BIGINT) long groupId,
+            @SqlType("array(integer)") Block groupingOrdinalsBlock,
+            @SqlType("array(array(integer))") Block groupingSetOrdinalsBlock)
+    {
+        List<Integer> groupingOrdinals = new ArrayList<>(groupingOrdinalsBlock.getPositionCount());
```
Comment (petroav) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/GroupingOperationFunction.java]: I think we should perform the extraction of `groupingOrdinalsBlock` and `groupingSetOrdinalsBlock` once and re-use the results for every subsequent invocation of the function. @raghavsethi mentioned that at runtime multiple threads can invoke the same function instance so we should make sure that access to and construction of `groupingOrdinals` and `groupingSetOrdinals` is properly synchronized. I was thinking of guarding the code block that extracts the values with a lock.

What do you think? Is this optimization worthwhile?

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/ExpressionAnalyzer.java [hunk 2]
```diff
@@ -1065,6 +1064,24 @@ private Type coerceToSingleType(StackableAstVisitorContext<AnalysisContext> cont
         }
     }
 
+    public static Signature resolveFunction(FunctionCall node, boolean isApproximate, List<TypeSignature> argumentTypes, FunctionRegistry functionRegistry)
```
Comment (petroav) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/ExpressionAnalyzer.java]: I extracted the function resolution code into a method so that I can re-use it when rewriting the `GroupingOperation` to a `FunctionCall` and ensure that the `grouping()` function exists and if it doesn't fail before execution.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/ExpressionInterpreter.java [hunk 3]
```diff
@@ -882,7 +882,7 @@ protected Object visitFunctionCall(FunctionCall node, Object context)
             }
 
             // do not optimize non-deterministic functions
-            if (optimize && (!function.isDeterministic() || hasUnresolvedValue(argumentValues))) {
+            if (optimize && (!function.isDeterministic() || hasUnresolvedValue(argumentValues) || node.getName().toString().equals("grouping"))) {
```
Comment (petroav) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/ExpressionInterpreter.java]: When creating the `FunctionCall` node I use a `LongLiteral` as a place holder for the `groupid` channel thus making all arguments to the function literals. Consequently, the `ExpressionInterpreter` attempts to compute and inline the function during planning. This can be prevented by (1) marking the function as non-deterministic or (2) using a `QualifiedNameReference` instead of the `LongLiteral` as the place holder for the `groupid` channel or (3) the code I added above.

(1) doesn't make sense because the function is deterministic, for the same inputs it will generate the same outputs. I tried (2) but ran into all sorts of trouble in `TranslationMap`. I think (3) is the most elegant solution.

**Comment by petroav** [comment]
Adding the `syntax-needs-review` label to get approval on the changes to the SQL grammar.

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/GroupingOperationFunction.java (hunk 1 above)
Comment (raghavsethi) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/GroupingOperationFunction.java]: I'm going to let @martint tell you exactly how it should be done, but I think it's definitely important to either use Java types or do some kind of optimization to prevent having to decode all the blocks for every row.

Comment (petroav) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/GroupingOperationFunction.java]: Yea that would be ideal. I looked around the code but couldn't figure it out. I'd be curious to know how it can be done.

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/GroupingOperationFunction.java [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+
+import static com.facebook.presto.spi.type.IntegerType.INTEGER;
+import static com.facebook.presto.spi.type.StandardTypes.BIGINT;
+
+public final class GroupingOperationFunction
+{
+    private GroupingOperationFunction() {}
+
+    @ScalarFunction
+    @Description("Returns a bitmap as an integer, indicating which columns are present in the grouping")
```
Comment (raghavsethi) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/GroupingOperationFunction.java]: AFAIK the description is used for publicly accessible functions, this isn't really a good fit for that. I don't think this requires a description.

Comment (raghavsethi) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/GroupingOperationFunction.java]: What *is* required, though, is documentation for what the arguments mean. The second and third arguments are especially confusing (and I wrote grouping sets!)

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/AggregationAnalyzer.java [hunk 5]
```diff
@@ -450,6 +454,23 @@ public Boolean visitRow(Row node, final Void context)
...
 
         @Override
+        public Boolean visitGroupingOperation(GroupingOperation node, Void context)
+        {
+            Collection<Expression> argumentsNotInGroupBy = node.getGroupingColumns().stream()
+                    .filter((argument) -> !expressions.contains(argument))
+                    .collect(Collectors.toList());
+            if (!argumentsNotInGroupBy.isEmpty()) {
+                throw new SemanticException(
+                        COLUMN_MISMATCH_GROUPING_AND_GROUPING_SET,
```
Comment (raghavsethi) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/AggregationAnalyzer.java]: I think invalid_function_argument would suffice

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/SemanticErrorCode.java [hunk 6]
```diff
@@ -80,4 +80,7 @@
     INVALID_PRIVILEGE,
 
     AMBIGUOUS_FUNCTION_CALL,
+
+    COLUMN_MISMATCH_GROUPING_AND_GROUPING_SET,
+    NO_CORRESPONDING_GROUPING_SET_ROLLUP_CUBE_TO_GROUPING,
```
Comment (raghavsethi) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/SemanticErrorCode.java]: Like I said, I think we can get by without these.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/ExpressionInterpreter.java [hunk 7]
```diff
@@ -1066,7 +1066,7 @@ protected Object visitRow(Row node, Object context)
                 return new Row(toExpressions(values, parameterTypes));
             }
             else {
-                BlockBuilder blockBuilder =  new InterleavedBlockBuilder(parameterTypes, new BlockBuilderStatus(), cardinality);
```
Comment (raghavsethi) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/ExpressionInterpreter.java]: Unrelated change

File: presto-main/src/main/java/com/facebook/presto/sql/planner/ExpressionInterpreter.java (hunk 3 above)
Comment (raghavsethi) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/ExpressionInterpreter.java]: AFAIK, non-deterministic also means that the function will not be pushed down. This would be a nice side-effect to have, but isn't the best way to achieve this.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/QueryPlanner.java [hunk 8]
```diff
@@ -269,14 +277,40 @@ private PlanBuilder filter(PlanBuilder subPlan, Expression predicate, Node node)
...
+    private PlanBuilder project(PlanBuilder subPlan, Iterable<Expression> expressions, Node node)
     {
         TranslationMap outputTranslations = new TranslationMap(subPlan.getRelationPlan(), analysis);
 
         ImmutableMap.Builder<Symbol, Expression> projections = ImmutableMap.builder();
         for (Expression expression : expressions) {
             Symbol symbol = symbolAllocator.newSymbol(expression, analysis.getTypeWithCoercions(expression));
 
+            if (expression instanceof GroupingOperation && node instanceof QuerySpecification) {
+                FunctionCall rewrittenExpression = subPlan.rewriteGroupingOperationToFunctionCall((GroupingOperation) expression, (QuerySpecification) node);
```
Comment (raghavsethi) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/QueryPlanner.java]: If it's going to stay like this, it should be factored out. I don't think it will though, this spraying of special cases all over the place is pretty hacky.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/SymbolToInputRewriter.java [hunk 9]
```diff
@@ -40,8 +44,22 @@ public SymbolToInputRewriter(Map<Symbol, Integer> symbolToChannelMapping)
     public Expression rewriteQualifiedNameReference(QualifiedNameReference node, Void context, ExpressionTreeRewriter<Void> treeRewriter)
     {
         Integer channel = symbolToChannelMapping.get(Symbol.fromQualifiedName(node.getName()));
-        Preconditions.checkArgument(channel != null, "Cannot resolve symbol %s", node.getName());
```
Comment (raghavsethi) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/SymbolToInputRewriter.java]: Unrelated change. Put it into a separate commit.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/ExpressionInterpreter.java (hunk 3 above)
Comment (petroav) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/ExpressionInterpreter.java]: Just to confirm, you'd like me to remove this check and instead mark the function as non-deterministic because the definition of non-deterministic includes functions that should not be inlined?

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/GroupingOperationFunction.java (hunk 4 above)
Comment (petroav) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/GroupingOperationFunction.java]: I removed the `@Description` annotation and added some Java Doc explaining the arguments and the function. Let me know what you think. I also changed the return type to `INTEGER`.

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/AggregationAnalyzer.java (hunk 5 above)
Comment (petroav) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/AggregationAnalyzer.java]: Done in efebbd90a82739fe067f2d86a92b1bde998086fd.

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/SemanticErrorCode.java (hunk 6 above)
Comment (petroav) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/SemanticErrorCode.java]: Removed both in efebbd90a82739fe067f2d86a92b1bde998086fd.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/ExpressionInterpreter.java (hunk 7 above)
Comment (petroav) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/ExpressionInterpreter.java]: Removed in efebbd90a82739fe067f2d86a92b1bde998086fd and will add back in a later commit.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/QueryPlanner.java (hunk 8 above)
Comment (petroav) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/QueryPlanner.java]: Extracted it into a method in efebbd90a82739fe067f2d86a92b1bde998086fd.

It's definitely a bit of a hack but I don't think there's a better way to do it without the new IR that @martint talked about.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/SymbolToInputRewriter.java (hunk 9 above)
Comment (petroav) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/SymbolToInputRewriter.java]: Remove in efebbd90a82739fe067f2d86a92b1bde998086fd and will add back in a later commit.

**Comment by petroav** [comment]
Addressed @raghavsethi's comments. @martint when you get a chance can you take a look and let me know how I can pass Java objects to the function and avoid the current decoding.

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/GroupingOperationFunction.java (hunk 1 above)
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/GroupingOperationFunction.java]: The only way to do that today is to create a synthetic type like we did with regexp patterns, introduce a functions that can transform the grouping ordinals and grouping sets arrays into such an object, and let the constant folder turn them into an object of that type before code generation.

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/GroupingOperationFunction.java [hunk 10]
```diff
@@ -0,0 +1,104 @@
...
+            for (int j = 0; j < subList.getPositionCount(); j++) {
+                ordinals.add(subList.getInt(j, 0));
+            }
+            groupingSetOrdinals.add(ordinals);
+        }
+
+        BitSet groupingBinary = new BitSet(groupingOrdinals.size());
+        groupingBinary.set(0, groupingOrdinals.size());
+        List<Integer> groupingSet = groupingSetOrdinals.get((int) groupId);
+        for (Integer groupingColumn : groupingSet) {
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/GroupingOperationFunction.java]: Can these be null?
